Apps/NeuroscanAcquireLSL/NeuroscanAcquireProtocol.h

```diff
@@ -0,0 +1,120 @@
+#ifndef NEUROSCANACQUIREPROTOCOL_H
+#define NEUROSCANACQUIREPROTOCOL_H
+
+#include <QByteArray>
+#include <QDataStream>
+#include <QDebug>
+#include <QtEndian>
+enum {
+    HeaderIdCtrl = 0x4354524c, // 'CTRL'
+    HeaderIdFile = 0x46494c45, // 'FILE'
+    HeaderIdData = 0x44415441, // 'DATA'
+
+    GeneralControlCode = 1,
+    RequestForVersion = 1,
+    ClosingUp,
+    ServerDisconnected,
+    ServerControlCode = 2,
+    StartAcquisition = 1,
+    StopAcquisition,
+    StartImpedance,
+    ChangeSetup,
+    DCCorrection,
+    ClientControlCode = 3,
+    RequestEDFHeader = 1,
+    RequestAstFile,
+    RequestStartData,
+    RequestStopData,
+    RequestBasicInfo,
+
+    SetupFile = 1,
+    NeuroscanASTFormat = 1,
+    CtfDSFormat = 0x43544644, // 'CTFD', BCI2000-specific extension
+
+    DataType_InfoBlock = 1,
+    InfoType_Version = 1,
+    InfoType_EdfHeader,
+    InfoType_BasicInfo,
+
+    DataType_EegData = 2,
+    DataTypeRaw16bit = 1,
+    DataTypeRaw32bit,
+};
+
+class NsaPacket {
+private:
+    qint32 mId;
+    qint16 mCode;
+    qint16 mValue;
+    qint32 mDataSize;
+    QByteArray buffer;
+
+public:
+    NsaPacket() : mId(0), mCode(0), mValue(0), mDataSize(0) {}
+    NsaPacket(qint32 id, qint16 code, qint16 value) {
+        mId = id;
+        mCode = code;
+        mValue = value;
+        mDataSize = 0;
+        buffer.clear();
+    }
+
+public:
+    int headerSize() { return 12; }
+    int dataSize() { return mDataSize; }
+    QByteArray& serialize() {
```

Comment:
Most implementations (especially the longer ones) should be in a separate cpp file